*add advanced store *reorganize sections *add icon,favicon (#53)
merantix-momentum/squirrel-core: docs/advanced/store.rst

Squirrel provides two serialization formats: :py:class:`MessagepackSerializer` and :py:class:`JsonSerializer`.
While JSONL might be preferable for interoperability or being human-readable, Messagepack is
faster to encode and decode and produces smaller files. Messegepack is the recommended format,
unless you have specific constraints or requirements. To demonstrate that Messagepack produces smaller files, we included the code
snippet below. We see that the files are around ~20% smaller compared to JSONL.

.. code-block:: python

  import numpy as np
  from squirrel.iterstream import IterableSource
  from squirrel.store import SquirrelStore
  from squirrel.serialization import MessagepackSerializer, JsonSerializer
    # creating random samples
    def get_sample(x):
        return {
            "img": np.random.random((20, 20, 3)),
            "label": x
        }

    N = 100_000

    summary = []

    for ser in [MessagepackSerializer, JsonSerializer]:
        tmpdir = tempfile.TemporaryDirectory()
        store = SquirrelStore(url=tmpdir.name, serializer=ser())

>         IterableSource(range(N)).map(get_sample).batched(1000).async_map(store.set).join() # asynch writing to store

Reviewer: ```suggestion
        IterableSource(range(N)).map(get_sample).batched(1000).async_map(store.set).join() # async writing to store
```

        size_mb = sum(f.stat().st_size for f in Path(tmpdir.name).glob('**/*') if f.is_file()) / 10e6 # total storage size in mb
        summary.append({"serializer": ser.__name__, "size_mb": size_mb})
        tmpdir.cleanup()
    print(summary)

Output::

    [{'serializer': 'MessagepackSerializer', 'size_mb': 90.6476465}, {'serializer': 'JsonSerializer', 'size_mb': 109.4487942}]

Sharding
--------------

> :py:class:`squirrel.store.SquirrelStore`  uses a concept called sharding (:ref:`usage/store:Sharding`) to efficiently store and load data.

Reviewer: ```suggestion
:py:class:`squirrel.store.SquirrelStore` uses a concept called sharding (:ref:`usage/store:Sharding`) to efficiently store and load data.
```

In simple terms, sharding simply means to divide and bundle your data into shards that are identified with a unique key.
There are several considerations for deciding appropriate shard size:
    1. Parallelizing read and write operation: the higher the number of shards, the bigger the opportunity for parallelizing read and write operations.Parallel write may be done with e.g. `Spark <https://github.com/merantix-momentum/squirrel-datasets-core/blob/main/examples/09.Spark_Preprocessing.ipynb/>`_
    2. Limit on the memory of the process when writing the shard. The :py:meth:`squirrel.store.store.set` accept a shard. This means that the whole shard has to be in memory for writing it. While technically shards of any size could have been created by incrementally writing to a single shard, we opted for this approach as it makes parallel and distributed write operations easier.

>     3. Randomizing during deep learning training: To shuffle the shards, one simply can shuffle the shard keys. That means, the more shards we have, the closer the shuffling process approaches a fully random shuffling. Internally there is another mechanism to shuffle samples within the shards (reference). However, the main idea of increasing the “degree” of randomness, with increasing shard counts remains.

Reviewer: Could you have one sentence per line here and also in general? Shorter lines are more readable

Custom Stores
--------------
:py:class:`squirrel.store.SquirrelStore` can handle most use cases, when you are reading and writing to a filesystem.
That includes all filesystems supported by `fsspec <https://filesystem-spec.readthedocs.io/en/latest/>`_ e.g google cloud storage or S3.

> However, if you have different use-cases such as reading data via HTTP-requests or retrieving from a database, you  may need to implement your own Store.
> The code snippet below implements a Store connecting a sqlite database. Here we can see that the concepts of sharding and serialization are not inherent to Store per se.

Reviewer: ```suggestion
However, if you have different use-cases such as reading data via HTTP requests or retrieving from a database, you may need to implement your own Store.
The code snippet below implements a Store connecting to a SQLite database. Here we can see that the concepts of sharding and serialization are not inherent to Store per se.
```

snippet below. We see that the files are around ~20% smaller compared to JSONL.

.. code-block:: python

  import numpy as np
  from squirrel.iterstream import IterableSource
  from squirrel.store import SquirrelStore
  from squirrel.serialization import MessagepackSerializer, JsonSerializer
    # creating random samples
    def get_sample(x):
        return {
            "img": np.random.random((20, 20, 3)),
            "label": x
        }

    N = 100_000

    summary = []

    for ser in [MessagepackSerializer, JsonSerializer]:
        tmpdir = tempfile.TemporaryDirectory()
        store = SquirrelStore(url=tmpdir.name, serializer=ser())
        IterableSource(range(N)).map(get_sample).batched(1000).async_map(store.set).join() # async writing to store
        size_mb = sum(f.stat().st_size for f in Path(tmpdir.name).glob('**/*') if f.is_file()) / 10e6 # total storage size in mb
        summary.append({"serializer": ser.__name__, "size_mb": size_mb})
        tmpdir.cleanup()
    print(summary)

Output::

    [{'serializer': 'MessagepackSerializer', 'size_mb': 90.6476465}, {'serializer': 'JsonSerializer', 'size_mb': 109.4487942}]

Sharding
--------------

> :py:class:`squirrel.store.SquirrelStore` uses a concept called sharding (:ref:`usage/store:Sharding`) to efficiently store and load data.

Reviewer: `usage/store:Sharding` is not a valid reference now, maybe the label/title has changed?

There are several considerations for deciding appropriate shard size:
    1. Parallelizing read and write operation: the higher the number of shards, the bigger the opportunity for parallelizing
    read and write operations.Parallel write may be done with e.g.
    `Spark <https://github.com/merantix-momentum/squirrel-datasets-core/blob/main/examples/09.Spark_Preprocessing.ipynb/>`_

    2. Limit on the memory of the process when writing the shard. The :py:meth:`squirrel.store.store.set` accept a shard.
    This means that the whole shard has to be in memory for writing it. While technically shards of any size could have
    been created by incrementally writing to a single shard, we opted for this approach as it makes parallel and distributed write operations easier.

    3. Randomizing during deep learning training: To shuffle the shards, one simply can shuffle the shard keys.
    That means, the more shards we have, the closer the shuffling process approaches a fully random shuffling.
    Internally there is another mechanism to shuffle samples within the shards (reference).

>     However, the main idea of increasing the “degree” of randomness, with increasing shard counts remains.

Reviewer: For numbered lists, I would just use `#.` notation so that we do not need to keep the count of bulletpoints ourselves. Also, there has to be an empty line before the bullet list. Then, for each bulletpoint, the texts should start at the same indentation level (e.g. for the first bulletpoint, "read and write operations ..." has to be aligned with "Parallelizing ...", not with "1.")

snippet below. We see that the files are around ~20% smaller compared to JSONL.

.. code-block:: python

  import numpy as np
  from squirrel.iterstream import IterableSource
  from squirrel.store import SquirrelStore
  from squirrel.serialization import MessagepackSerializer, JsonSerializer
    # creating random samples
    def get_sample(x):
        return {
            "img": np.random.random((20, 20, 3)),
            "label": x
        }

    N = 100_000

    summary = []

    for ser in [MessagepackSerializer, JsonSerializer]:
        tmpdir = tempfile.TemporaryDirectory()
        store = SquirrelStore(url=tmpdir.name, serializer=ser())
        IterableSource(range(N)).map(get_sample).batched(1000).async_map(store.set).join() # async writing to store
        size_mb = sum(f.stat().st_size for f in Path(tmpdir.name).glob('**/*') if f.is_file()) / 10e6 # total storage size in mb
        summary.append({"serializer": ser.__name__, "size_mb": size_mb})
        tmpdir.cleanup()
    print(summary)

Output::

    [{'serializer': 'MessagepackSerializer', 'size_mb': 90.6476465}, {'serializer': 'JsonSerializer', 'size_mb': 109.4487942}]

> Sharding

Reviewer 2: describing what shards and samples are should be in the basic usage as it concerns all users. The considerations may live here. 

I would suggest to provide a simple code snippets to illustrate a sample and a shard, and also pointing out to `squirrel.constants.SampleType` and `squirrel.constants.ShardType`

Shards stores a predetermined amount of samples given through the shard size in a fixed order. Each shard is then identified through a unique key.
There are several considerations for deciding appropriate shard size:

    #. Parallelizing read and write operation: the higher the number of shards, the bigger the opportunity for parallelizing
       read and write operations.Parallel write may be done with e.g.
       `Spark <https://github.com/merantix-momentum/squirrel-datasets-core/blob/main/examples/09.Spark_Preprocessing.ipynb/>`_

    #. Limit on the memory of the process when writing the shard. The :py:meth:`squirrel.store.store.set` accept a shard.
       This means that the whole shard has to be in memory for writing it. While technically shards of any size could have
       been created by incrementally writing to a single shard, we opted for this approach as it makes parallel and distributed write operations easier.

    #. Randomizing during deep learning training: To shuffle the shards, one simply can shuffle the shard keys.
       That means, the more shards we have, the closer the shuffling process approaches a fully random shuffling.

>        Internally there is another mechanism to shuffle samples within the shards (reference).

Reviewer 2: If this refer to the `.shuffle()` method of `Composable`, it actually shuffles samples on the stream.

Author: Then I'll reference to .shuffle(), if that's fine for you

       However, the main idea of increasing the “degree” of randomness, with increasing shard counts remains.

Custom Stores
--------------

> However, if you have different use-cases such as reading data via HTTP requests or retrieving from a database, you may need to implement your own Store.

Reviewer 2: ```suggestion
:py:class:`AbstractStore` defines an abstraction to provide a key/value api on top of any storage. It may optionally use :py:class:`SquirrelSerializer`. If you have a specific use-case which is not natively provided such as reading data via HTTP requests or retrieving from a database, you may need to implement your own Store.
```

The code snippet below implements a Store connecting to a SQLite database.
Here we can see that the concepts of sharding and serialization are not inherent to Store per se.

> .. code-block:: python

Reviewer 2: no strong opinion, but seems to me that this fits better as a notebook tutorial which is then linked here.

Author: True, but the integration of notebooks is an open discussion, right?

Store
=====
Architecture
____________
In the UML diagram below we show how the SquirrelStore connects to its abstract class and different serializers.

.. mermaid::

    classDiagram

        AbstractStore <|-- SquirrelStore
        SquirrelStore *-- SquirrelSerializer
        SquirrelSerializer <|-- MessagepackSerializer
        SquirrelSerializer <|-- JSONSerializer
        <<abstract>> AbstractStore
        class AbstractStore {

            set(key, value)
            get(key) Iterable~Dict~
            keys() Iterable~Any~
        }

       class SquirrelStore {
           serializer: SquirrelSerializer

           set(key, value)
           get(key) Iterable~Dict~
           keys() Iterable~str~
       }
        <<abstract>> SquirrelSerializer
       class SquirrelSerializer {

        serialize(obj)
        deserialize(obj)
       }

       class MessagepackSerializer {

           serialize(obj) bytes
           deserialize(bytes) Any
       }

       class JSONSerializer {

           serialize(obj) bytes
           deserialize(bytes) Any
       }

Serialization
--------------

> :py:class:`squirrel.store.SquirrelStore` uses a Serializer to store shards of samples as singular files onto the storage backend (e.g. filesytem, object store, etc.).

Reviewer: ```suggestion
:py:class:`squirrel.store.SquirrelStore` uses a Serializer to store shards of samples as singular files onto the storage backend (e.g. filesystem, object store, etc.).
```

> Squirrel provides two serializers formats: :py:class:`MessagepackSerializer` and :py:class:`JsonSerializer`.

Reviewer: ```suggestion
Squirrel provides two serializers: :py:class:`MessagepackSerializer` and :py:class:`JsonSerializer`.
```

While JSONL might be preferable for interoperability or being human-readable, Messagepack is

> faster to encode and decode and produces smaller files. Messegepack is the recommended format,

Reviewer: ```suggestion
faster to encode and decode and produces smaller files. Messagepack is the recommended format,
```

> unless you have specific constraints or requirements. To demonstrate that Messagepack produces smaller files, we included the code

Reviewer: ```suggestion
unless you have specific constraints or requirements. To demonstrate that Messagepack produces smaller files, we include the code
```

       read and write operations.Parallel write may be done with e.g.
       `Spark <https://github.com/merantix-momentum/squirrel-datasets-core/blob/main/examples/09.Spark_Preprocessing.ipynb/>`_

    #. Limit on the memory of the process when writing the shard. The :py:meth:`squirrel.store.Store.set` accept a shard.
       This means that the whole shard has to be in memory for writing it. While technically shards of any size could have
       been created by incrementally writing to a single shard, we opted for this approach as it makes parallel and distributed write operations easier.

    #. Randomizing during deep learning training: To shuffle the shards, one simply can shuffle the shard keys.
       That means, the more shards we have, the closer the shuffling process approaches a fully random shuffling.
       Internally there is another mechanism to shuffle samples within the shards (:py:meth:`squirrel.base.Composable.shuffle`).
       However, the main idea of increasing the “degree” of randomness, with increasing shard counts remains.

Custom Stores
--------------

> :py:class:`AbstractStore` defines an abstraction to provide a key/value api on top of any storage.

Reviewer: ```suggestion
:py:class:`AbstractStore` defines an abstraction to provide a key/value API on top of any storage.
All stores are expected to obey this API.
```

      return hash(value), value

    class SQLiteStore(AbstractStore):
      def __init__(self, db_name:str):
        self._con = sqlite3.connect(db_name)
        self._cur = self._con.cursor()
        self._cur.execute("DROP TABLE IF EXISTS demo") # drop existing table and create a simple key-value table
        self._cur.execute("""CREATE TABLE demo (key INTEGER PRIMARY KEY, value TEXT);""")
        self._con.commit()

      def set(self, key: t.Any, value: t.Any) -> None:
        # insert value given a key
        self._cur.execute("INSERT INTO demo VALUES (?,?)", (key, value))
        self._con.commit()

      def get(self, key: t.Any) -> t.Iterable:
        # retrieve value with the key
        value = self._cur.execute("SELECT value FROM demo WHERE key=?", key).fetchall()
        return value

      def keys(self) -> t.Iterable:
        # return all the keys stored
        return self._cur.execute("SELECT key FROM demo")

      def close(self):
        # close the database connection
        self._con.close()

    # We create the SQLite db and insert key-value pairs into it
    N = 100_000
    with tempfile.TemporaryDirectory() as temp_dir:

>       store = SQLiteStore('temp.db')

Author: works!

       read and write operations.Parallel write may be done with e.g.
       `Spark <https://github.com/merantix-momentum/squirrel-datasets-core/blob/main/examples/09.Spark_Preprocessing.ipynb/>`_

    #. Limit on the memory of the process when writing the shard. The :py:meth:`squirrel.store.Store.set` accept a shard.
       This means that the whole shard has to be in memory for writing it. While technically shards of any size could have
       been created by incrementally writing to a single shard, we opted for this approach as it makes parallel and distributed write operations easier.

    #. Randomizing during deep learning training: To shuffle the shards, one simply can shuffle the shard keys.
       That means, the more shards we have, the closer the shuffling process approaches a fully random shuffling.
       Internally there is another mechanism to shuffle samples within the shards (:py:meth:`squirrel.base.Composable.shuffle`).
       However, the main idea of increasing the “degree” of randomness, with increasing shard counts remains.

Custom Stores
--------------
:py:class:`AbstractStore` defines an abstraction to provide a key/value API on top of any storage.

> All stores are expected to obey this API.

Reviewer 2: ```suggestion
All stores should conform to this abstraction.
```
